### ISSUE TITLE
fastboot compatibility

### ISSUE THREAD

**Issue by bantic** [yapplabs/ember-wormhole#47]
Is there any plan for or interest in adding fastboot compatibility to ember-wormhole?

I've made a [proof-of-concept here](https://github.com/bantic/ember-wormhole/commit/f3c7169a318b9f838bbddbe92d5ccb97b79f394e) that uses a few private APIs to work around Fastboot's restrictions and make it work in Fastboot. If there's interest in adding this functionality to ember-wormhole there are probably a few safer/more-public APIs that can be used instead.

Another idea suggested by @mixonic:

 *  alt idea is that helpers can return nodes, so add helper calls before and after the yield statement, have them push the node they return onto an array shared with the parent component)
 * parent component uses afterRender when that array is updated w/ the nodes, and uses them to walk instead of trying to find the first/last child nodes

**Comment by stefanpenner** [comment]
@bantic I would say yes

**Comment by bantic** [comment]
@stefanpenner 👍  If you have a chance to give any feedback on the methods used in [this POC](https://github.com/bantic/ember-wormhole/commit/f3c7169a318b9f838bbddbe92d5ccb97b79f394e) or the alt idea from @mixonic above I'd be happy to work toward one of them.

**Comment by stefanpenner** [comment]
@bantic if you remind me this weekend, I absolutely can.

**Comment by mixonic** [comment]
I think there are two thing that *could* change in @bantic's spike

* SimpleDOM doesn't support selectors (like `getElementById`) and will not. So the PoC creates a custom selector implementation. This can be removed if we permit an element to be passed to wormholes instead of an id. `{{ember-wormhole to=someDOMNode}}`. When using Fastboot, this would just be a SimpleDOM node which should mean everything Just Works. It would also allow for scoping the targets of wormholes more locally. That all wormholes are in a global namespace (document node ids) is a little unfortunate, and I've needed to generate unique IDs for wormhole before where I could have simply passed an element.
* Second, reaching into `this._renderNode.firstNode` is mildly unfortunate as it is private API. It is a dance, but I think it is possible to do this with public API. Some thing like the following.

```hbs
{{! ember-wormhole.hbs }}
{{wormhole-head registry=registry as='head'}}
{{yield}}
{{wormhole-tail registry=registry as='tail'}}
```

Then in JS:

```js
// wormhole-tail
export default Ember.Helper.helper(([registry, slot]) => {
  let node = createTextNode(); // must be DOM and fastboot compatible
  registry[slot] = node;
  return node;
});
```

```js
// ember-wormhole.js
/* ...snip... */
  init() {
    this._super();
    this.registry = {};
  }
/* ...snip... */
  willRender() { // willRender runs in Fastboot
    run.schedule('afterRender', () => {
      this._firstNode = this.registry.head;
      this._lastNode = this.registry.tail;
      this.appendToDestination();
    });
  }
/* ...snip... */
```

all public API

**Comment by bantic** [comment]
@stefanpenner Ping, since you had asked for a reminder this weekend to look this one over. Matt's approach above looks good to me. For reference the working POC I have is here: https://github.com/bantic/ember-wormhole/commit/f3c7169a318b9f838bbddbe92d5ccb97b79f394e

**Comment by bantic** [comment]
Here's another attempt using a helper named `ember-wormhole-placeholder` as described above by @mixonic. https://github.com/yapplabs/ember-wormhole/compare/master...bantic:fastboot-without-private-apis

**Comment by bantic** [comment]
There are a few blockers I'd like to get some feedback on:

 * Supported versions of Ember

ember-wormhole includes support for ember back to 1.10.1, which predates many current Ember features. The approach I am currently working on uses an Ember helper, but `Ember.Helper.helper` isn't available until Ember 1.13. Fastboot only supports Ember versions 2.3+, anyway, but it would be best if the fastboot-related updates to ember-wormhole work with all the versions of Ember it supports, rather than having drastically different code for fastboot-compatible ember-wormhole versus normal ember-wormhole. It seems possible to do this if ember-wormhole is changed to only support ember 1.13+.

Would the ember-wormhole maintainers consider dropping pre-Ember 1.13 support in a semver-breaking version update?

 * Usage of ember-wormhole by extending the component directly

The approach I'm taking to make wormhole work in Fastboot is to add a layout to the component that includes a helper, e.g.:
```
// ember-wormhole.hbs
{{ember-wormhole-placeholder as="firstNode"}}
{{yield}}
{{ember-wormhole-placeholder as="lastNode"}}
```

If a consumer imports the ember-wormhole component and extends it using a different layout, this overrides the built-in one, e.g.:

```
// components/x-favicon.js
import Wormhole from 'ember-wormhole/components/ember-wormhole';
export default Wormhole.extend({
  // ...
});

// templates/components/x-favicon.hbs
... {{! this template overrides the one at ember-wormhole.hbs}}
```

This usage isn't described in the README anywhere but it happens in the dummy app in tests [in the x-favicon component](https://github.com/yapplabs/ember-wormhole/blob/master/tests/dummy/app/components/x-favicon.js)).

A [code search for "import ember-wormhole/components/ember-wormhole" on github](https://github.com/search?p=1&q=import+ember-wormhole%2Fcomponents%2Fember-wormhole&ref=searchresults&type=Code&utf8=%E2%9C%93) only reveals a few places where wormhole is used in this way, and from what I can tell the only place that it is used in a way that would be broken by the changes I'm attempting is [here](https://github.com/chrislopresto/ember-reveal-js/blob/915ef469771d2a3424f4efc267c4fb824d7f9444/addon/components/reveal-presentation/component.js), because that component extends `ember-wormhole` and also brings its own template/layout.

Is this extension-based usage intended to be supported, or could it be dropped in future versions of ember-wormhole? If it is intended, it's probably not possible to support Fastboot by adding an ember-wormhole template as shown above, and the best path forward for will likely require more private APIs (as in my earlier spike, [mentioned above](https://github.com/yapplabs/ember-wormhole/issues/47#issue-154521433)).

**Comment by lukemelia** [comment]
@bantic Thanks for the thorough writeup.

Regarding supported versions of Ember, I think it is reasonable at this point to drop support for ember < 1.13 with an appropriate semver bump.

Regarding the wormhole subclassing, I think the pattern for the type of usage you described should be composition rather than inheritance. Wormhole should be open for extension via inheritance, but I think it is OK for there to be constraints on what the template needs to have in order to function as expected.

I would love to get @krisselden's perspective on this thread as well.

**Comment by chrislopresto** [comment]
Picking back up on one of @mixonic's points. I think wormhole should support being passed an element as opposed to just an id selector. We could add support for this in a minor release and then take advantage of it here if it simplifies FastBoot matters.

I agree with @lukemelia about dropping support for < 1.13 with a semver bump.

Constraints on subclassed templates also seem fine. As an added bonus, subclassers could trick out their placeholder components... maybe add some sweet tail fins.

**Comment by krisselden** [comment]
I think in general, extending a class requires knowledge of how it works, instead of knowledge of input/output.  I think it is fine for a power user but I've never been fan of SemVer guarantees about extending things via overrides, it pretty much shuts down any improvements to how things are accomplished.

**Comment by bantic** [comment]
Fixed by #54.
